osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -0,0 +1,399 @@
+            {
+                var timingPoint = group.ControlPoints.OfType<TimingControlPoint>().FirstOrDefault();
+                var difficultyPoint = beatmap.ControlPointInfo.DifficultyPointAt(group.Time);
+                var samplePoint = beatmap.ControlPointInfo.SamplePointAt(group.Time);
+                var effectPoint = beatmap.ControlPointInfo.EffectPointAt(group.Time);
+
+                // Convert beat length the legacy format
+                double beatLength;
+                if (timingPoint != null)
+                    beatLength = timingPoint.BeatLength;
+                else
+                    beatLength = -100 / difficultyPoint.SpeedMultiplier;
+
+                // Apply the control point to a hit sample to uncover legacy properties (e.g. suffix)
+                HitSampleInfo tempHitSample = samplePoint.ApplyTo(new HitSampleInfo());
+
+                // Convert effect flags to the legacy format
+                LegacyEffectFlags effectFlags = LegacyEffectFlags.None;
+                if (effectPoint.KiaiMode)
+                    effectFlags |= LegacyEffectFlags.Kiai;
+                if (effectPoint.OmitFirstBarLine)
+                    effectFlags |= LegacyEffectFlags.OmitFirstBarLine;
+
+                writer.Write(FormattableString.Invariant($"{group.Time},"));
+                writer.Write(FormattableString.Invariant($"{beatLength},"));
+                writer.Write(FormattableString.Invariant($"{(int)beatmap.ControlPointInfo.TimingPointAt(group.Time).TimeSignature},"));
+                writer.Write(FormattableString.Invariant($"{(int)toLegacySampleBank(tempHitSample.Bank)},"));
+                writer.Write(FormattableString.Invariant($"{toLegacyCustomSampleBank(tempHitSample.Suffix)},"));
+                writer.Write(FormattableString.Invariant($"{tempHitSample.Volume},"));
+                writer.Write(FormattableString.Invariant($"{(timingPoint != null ? "1" : "0")},"));
+                writer.Write(FormattableString.Invariant($"{(int)effectFlags}"));
+                writer.Write("\n");
+            }
+        }
+
+        private void handleHitObjects(TextWriter writer)
+        {
+            if (beatmap.HitObjects.Count == 0)
+                return;
+
+            writer.WriteLine("[HitObjects]");
+
+            foreach (var h in beatmap.HitObjects)
+            {
+                switch (beatmap.BeatmapInfo.RulesetID)
+                {
+                    case 0:
+                        handleOsuHitObject(writer, h);
+                        break;
+
+                    case 1:
+                        handleTaikoHitObject(writer, h);
+                        break;
+
+                    case 2:
+                        handleCatchHitObject(writer, h);
+                        break;
+
+                    case 3:
+                        handleManiaHitObject(writer, h);
+                        break;
+                }
+            }
```

Comment:
This reads weird to me with the switch inside of the loop instead of the opposite. It'll most likely have no effect on anything, I'm guessing not even performance (due to branch prediction), but it pops out as a strange construct to me.